### DISCUSSION TITLE
Shareth using the Crown of Imiriel

### DISCUSSION THREAD

**Discussion by IcemarkUK** [ChilliHugger/The-Lords-Of-Midnight#165] [Code Algorithm Questions] [answered]
In response to https://github.com/ChilliHugger/The-Lords-Of-Midnight/discussions/73#discussioncomment-5561869

TLDR; The code is technically correct because no one can use the Crown of Imiriel except Shareth, but AI lords do not use their objects. 

The crowns should all work the same however for completeness.

This is the code that checks if an object can be used by the character.
They have to be carrying THEIR object ( ObjectNo == CharacterNo )  and have not used THEIR object (BIT 7 of flags).

```
L7C60   CALL GetCharFlag_1                  ;
    BIT  7,A                                ;has the character
    JR   NZ,L7C78                           ;used his object?
    CALL GetCharObject                      ;is he carrying anything?
    CP   Nothing
    JR   Z,L7C78
    LD   B,A                                ;
    LD   A,(CharacterNumber)                ;Is it his special Object?
    CP   B                                  ;
    JR   NZ,L7C78                           ;
    LD   (SomeThingToUse),A                 ;If all met. Then something to Use.
```

Code called in response to pressing 'Use' in the UI. Sets Used flag and then Uses the object.

```
Use  ;L879E
    LD   A,(SomeThingToUse)
    CP   Nothing
    JP   Z,PollGameKeys
    CALL GetCharFlag_1                      ;
    SET  7,A                                ;Used Object
    CALL PutCharFlag_1                      ;
    CALL UseTheObject
    JP   CheckPlace
```

Called from here...

```
CheckChoicesKeys  ;L8654
    CALL GetCharTimeOfDay
    CP   Nightt
    JP   Z,PollGameKeys
    CALL CheckKeyCase
    CALL SelectCase
    DEFB #09
    CASE <'1',Rest>
    CASE <'2',Take>
    CASE <'3',Give>
    CASE <'4',EnterTunnel>
    CASE <'5',Recruit>
    CASE <'6',Post>
    CASE <'7',Attack>
    CASE <'8',Approach>
    CASE <'9',Use>
    JP   PollGameKeys
```

Object will now be used. 
Special objects 0, 1, 2, 3, 4
Crowns are 0, 1, and 4

```
UseTheObject  ;L67AB
    LD   A,(SomeThingToUse)
    CP   #05                                ;One of the special
    JR   C,L67BA                            ;objects?
    LD   A,FullUp                           ;
    CALL PutCharTired                       ;Reset These if any Other
    JP   PutCharDespondent
L67BA   CP   #02                            ;Spell Of Swiftness
    JR   Z,DealWithSpell
    CP   #03                                ;Runes of Protection
    JR   Z,DealWithRunes

; use the crowns
; brings all lords loyal to Luxor to the Curren location
; placing them in or removing them from the tunnel if need be
;
    CALL ClearMapOfArmys                    ;Crowns of Persuassion
    XOR  A
L67C6   LD   (CharacterNumber),A            ;Character No.
    CALL GetCharLoyalty                     ;is he loyal?
    CP   LoyalToMoonprince                  ;
    JR   NZ,L67F0                           ;Skip If Not.
    LD   BC,(CurrentLocation)               ;
    CALL PutCharLocation                    ;Move him To this Location
    LD   A,(CharIsInTheTunnel)
    CP   #00
    JR   Z,L67E8                            ;
    CALL GetCharFlag_1                      ;
    SET  6,A                                ;This lot basically puts
    CALL PutCharFlag_1                      ;the character in or out
    JR   L67F0                              ;of a tunnel if need be.
L67E8   CALL GetCharFlag_1                  ;
    RES  6,A                                ;
    CALL PutCharFlag_1
L67F0   LD   A,(CharacterNumber)
    INC  A
    CP   #80
    JR   NZ,L67C6                           ;Loop until finished.
    JP   AddArmysToMap
```

During night processing AI lords will take any object without consideration to special or not. Once they are carrying an object they will not take another - note: Fix required for this!

**Comment by IcemarkUK** [comment]
2 potential issues raised - #166 and #167

**Comment by FrozenEmpire** [comment, marked as answer]
Thanks Chris,

So the key learning for me is that AI characters never “Use” their objects, it is only a UI command.
Which kind of makes the Crown of Imiriel “for decoration only”

Or maybe that can be a custom difficulty option where she “thinks carefully when to use it” and then all those loyal to the heartstealer get transported to Shareth :)
